alibi_detect/od/base.py

```diff
@@ -1,32 +1,76 @@
-from __future__ import annotations
-from typing import Union
+from alibi_detect.utils.missing_optional_dependency import import_optional
 
-from typing_extensions import Protocol, runtime_checkable
+from typing import Union
+from typing_extensions import Literal, Protocol, runtime_checkable
 
 
 # Use Protocols instead of base classes for the backend associated objects. This is a bit more flexible and allows us to
 # avoid the torch/tensorflow imports in the base class.
 @runtime_checkable
 class TransformProtocol(Protocol):
-    """Protocol for transformer objects."""
-    def transform(self, x):
-        pass
+    """Protocol for transformer objects.
 
-    def _transform(self, x):
+    The :py:obj:`~alibi_detect.od.pytorch.ensemble.BaseTransformTorch` object provides abstract methods for
```

Comment:
The rendered link in docs doesn't work (works for the next one down): https://seldon--677.org.readthedocs.build/projects/alibi-detect/en/677/api/alibi_detect.od.base.html

Reply:
Ah, yep. I removed the `BaseFittedTransform` as I had wanted to use the private implementation pattern to fit the transform but I couldn't get autodoc to document the private methods in that case:

```py
class FitMixin:
    fitted = False

   def fit(self, x_ref):
       self._fit(self, x_ref)
       self.fitted = True

   def _fit(self, x_ref):
       raise NotImplemented()
```

The above requires that the actual fit implementation be done in `_fit` on each of the transformers that uses the Mixin and so those docstrings don't display in that case. We can obviously toggle `autodoc` to display the docstrings for `_fit` but that seems risky and perhaps will lead to users using `_fit` instead of `fit`. What we really want is the docstring for `fit` on the child class to be the docstring for `_fit` on the child class, but there didn't seem to be a way of doing that.

Reply:
I've changed the docstring above to point to both the `BaseTransformTorch` object and the `FitMixinTorch` as it models the interface for the combination of both.